[Cocoa][MSE] Seeking to the end of X.com videos will never complete

jernoble opened the pull request:
#### e3ec66bc9646bfbd69446c491205765c530e8989
<pre>
[Cocoa][MSE] Seeking to the end of X.com videos will never complete
<a href="https://bugs.webkit.org/show_bug.cgi?id=285776">https://bugs.webkit.org/show_bug.cgi?id=285776</a>
<a href="https://rdar.apple.com/136845120">rdar://136845120</a>

Reviewed by Jean-Yves Avenard.

We already allow selections immediately before a buffered range to succeed; also
apply the same &quot;fudge factor&quot; to allow seeking just past the end of a buffered range.

* LayoutTests/media/media-source-seek-past-end-expected.txt: Added.
* LayoutTests/media/media-source-seek-past-end.html: Added.
* Source/WebCore/platform/graphics/TrackBuffer.cpp:

Canonical link: <a href="https://commits.webkit.org/288962@main">https://commits.webkit.org/288962@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/912e52172e5fdede08fd90a4fcdcb5b6b554e759

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84691 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4416 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39079 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89830 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35742 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86776 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4505 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12391 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65916 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23744 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87736 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3427 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76968 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46189 "Found 1 new API test failure: /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/service-worker-registrations (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3305 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31180 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed layout tests; 22 flakes 2 failures; Uploaded test results; 21 flakes 3 failures; Compiled WebKit (warnings); layout-tests; Passed layout tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34817 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74169 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31985 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91206 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12030 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8791 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74392 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12257 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72772 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73517 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18234 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17888 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16331 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3478 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11982 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17422 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11816 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15310 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13562 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -978,7 +978,8 @@ - (void)removeAllVideoTargets;
     if (needsFlush == NeedsFlush::Yes)
         flushTrack(trackId);
-    if (trackBuffer.reenqueueMediaForTime(time, timeFudgeFactor()))
+    bool isEnded = m_loadFinished && !m_pendingAppends;
```
I think this won't be sufficient for webm; `Ref<GenericPromise> MediaPlayerPrivateWebM::seekTo(const MediaTime& time)` will not complete the seek if it's not within the buffered range, and the duration won't be.

If we want to handle this case, needs to be done there.

jyavenard commented on Source/WebCore/platform/graphics/TrackBuffer.cpp:
```diff
@@ -100,22 +100,35 @@ bool TrackBuffer::updateMinimumUpcomingPresentationTime()
+    if (currentSamplePTSIterator == m_samples.presentationOrder().end()) {
+        auto nextSampleIterator = m_samples.presentationOrder().findSampleStartingOnOrAfterPresentationTime(time);
+        if ((nextSampleIterator->first - time) <= timeFudgeFactor)
+            currentSamplePTSIterator = nextSampleIterator;
```
Is this needed here?

For the issue at and, below is all necessary right?

jernoble replied:
This is just refactoring the existing check.

jyavenard commented on LayoutTests/media/media-source/media-source-seek-past-end.html:
```diff
@@ -0,0 +1,77 @@
+
+        consoleWrite('-');
+        consoleWrite('Seek into the unbuffered video range');
+        run('video.currentTime = video.duration');
```
Would be nice if the test checked that seeking past the duration won't complete if the source isn't ended.

jernoble committed:
[Cocoa][MSE] Seeking to the end of X.com videos will never complete
https://bugs.webkit.org/show_bug.cgi?id=285776
rdar://136845120

Reviewed by Jean-Yves Avenard.

We already allow selections immediately before a buffered range to succeed; also
apply the same "fudge factor" to allow seeking just past the end of a buffered range.

* LayoutTests/media/media-source-seek-past-end-expected.txt: Added.
* LayoutTests/media/media-source-seek-past-end.html: Added.
* Source/WebCore/platform/graphics/TrackBuffer.cpp:

Canonical link: https://commits.webkit.org/288962@main
